### ISSUE TITLE
[BUG] calling List.sort triggers a "TypeError: r is not a function" (JS error)

### ISSUE THREAD

**Issue by sebferre** [ocsigen/js_of_ocaml#1007]
**Describe the bug**
When calling function List.sort from the standard library, a JS error is triggered.
It displays as:  TypeError: r is not a function
(the name of the function varies)

A call as simple as the following is enough.
List.sort (fun x y -> compare y x) [4;2;5;2;1]

The bug disappears when I call my own naive implementation of List.sort (Quicksort).

**Versions**
Version of packages used to reproduce the bug
ocamlc 4.09.1
js_of_ocaml 3.5.2
opam 2.0.6

**Comment by hhugo** [comment]
Please provide more context, I cannot reproduce the issue.

**Comment by sebferre** [comment]
I don't know what to add really.

I use packages yojson, ppx_deriving_yojson, csv in addition to js_of_ocaml and js_of_ocaml-lwt,
but the relevant part of the code makes use of none, it's basic OCaml.
The function in which the problem occurs is a match over a variant type, and in one case of the
match, I need to sort a list, and I call List.sort. 
At that position in the code, if I insert the simple call given in my initial post,
the problem occurs too, so it's not related to the other details of what I do in my function.
However, when I put the call to List.sort as a top 'let' statement, the problem does not occur.

**Comment by sebferre** [comment]
I did a number of checks, and it seems that the problem occurs in recursive constructors of my
variant type. Here is my variant type :

type expr = ...
and flower =
  | Return of expr (* NO bug *)
  | FileData of string * data * flower (* BUG *)
  | For of var * expr * bool * flower (* BUG *)
  | FLet of var * expr * flower
  | Where of expr * flower (* BUG *)
  | GroupBy of var list * flower
  | Project of var list * flower (* BUG *)
  | Slice of int * int option * flower
  | OrderBy of (expr * order) list * flower (* BUG *)
  | FConcat of flower list
  | FIf of flower * flower * flower
			       [@@deriving yojson]

**Comment by hhugo** [comment]
Can you share some code and instructions to reproduce the issue ?

**Comment by hhugo** [comment]
I'm guessing you run your program inside a web-browser.

Few questions:

How do you call js_of_ocaml ? are you passing any flags ? are you doing whole program compilation or separate compilation ?
Do you use custom javascript stubs ?

> It displays as: TypeError: r is not a function

You mean it displays in the javascript console ?

When debugging, I would suggest to use `--pretty` when calling js_of_ocaml.

**Comment by sebferre** [comment]
Yes, I run it inside a Web browser (firefox).
I call js_of_ocaml without any option. Here are the makefile-generated commands:

ocamlfind ocamlc -package yojson,ppx_deriving_yojson,csv -I ../../../lib -I ../../core -I ../../core/webapp -package js_of_ocaml,js_of_ocaml-lwt -linkpkg -o html/script.byte bintree.cmo myseq.cmo focus.cmo syntax.cmo lis.cmo jsutils.cmo webapp.cmo html.cmo widget_focus.cmo widget_suggestions.cmo widget_table.cmo jsoniq.cmo jsoniq_focus.cmo jsoniq_semantics.cmo jsoniq_syntax.cmo jsoniq_suggestions.cmo jsoniq_files.cmo jsoniq_lis.cmo jsoniq_webapp.ml
js_of_ocaml html/script.byte

Some modules are compiled without dependency of js_of_ocaml, and I want to keep it that way as
they are used in other projects.

The error is shown into an alert box because I catch the exception, and call the window##alert method on its textual repr.

**Comment by hhugo** [comment]
If you catch the error your-self try to catch `Js_of_ocaml.Js.Error` and log it in the console. Together with `--pretty` it should give you a backtrace.

```ocaml
| Js_of_ocaml.Js.Error js_error -> Firebug.console##error js_error
```

**Comment by sebferre** [comment]
Here is the backtrace (I can send you the script.js but it's ~19k lines)

TypeError: _U5_ is not a function
Stack trace:
_t3_/_Sd_</<@file:///local/ferre/prog/ocaml/fablis/instances/jsoniq/html/script.js:14890:30
_t3_@file:///local/ferre/prog/ocaml/fablis/instances/jsoniq/html/script.js:14954:30
_t2_@file:///local/ferre/prog/ocaml/fablis/instances/jsoniq/html/script.js:15244:25
_xX_/<@file:///local/ferre/prog/ocaml/fablis/instances/jsoniq/html/script.js:18157:24
caml_call3@file:///local/ferre/prog/ocaml/fablis/instances/jsoniq/html/script.js:2216:28
_yD_@file:///local/ferre/prog/ocaml/fablis/instances/jsoniq/html/script.js:19178:20
caml_call2@file:///local/ferre/prog/ocaml/fablis/instances/jsoniq/html/script.js:2214:28
_ahY_@file:///local/ferre/prog/ocaml/fablis/instances/jsoniq/html/script.js:12509:53
_ahZ_@file:///local/ferre/prog/ocaml/fablis/instances/jsoniq/html/script.js:12512:28
caml_call2@file:///local/ferre/prog/ocaml/fablis/instances/jsoniq/html/script.js:2214:28
_zk_/<@file:///local/ferre/prog/ocaml/fablis/instances/jsoniq/html/script.js:19176:41
caml_call1@file:///local/ferre/prog/ocaml/fablis/instances/jsoniq/html/script.js:2212:28
_adN_@file:///local/ferre/prog/ocaml/fablis/instances/jsoniq/html/script.js:13020:26
_adV_/</<@file:///local/ferre/prog/ocaml/fablis/instances/jsoniq/html/script.js:13137:79
caml_call1@file:///local/ferre/prog/ocaml/fablis/instances/jsoniq/html/script.js:2212:28
_alf_@file:///local/ferre/prog/ocaml/fablis/instances/jsoniq/html/script.js:12124:23
_alj_@file:///local/ferre/prog/ocaml/fablis/instances/jsoniq/html/script.js:12126:46
caml_call1@file:///local/ferre/prog/ocaml/fablis/instances/jsoniq/html/script.js:2212:28
_kJ_@file:///local/ferre/prog/ocaml/fablis/instances/jsoniq/html/script.js:11198:26
_alg_@file:///local/ferre/prog/ocaml/fablis/instances/jsoniq/html/script.js:12128:24
caml_call1@file:///local/ferre/prog/ocaml/fablis/instances/jsoniq/html/script.js:2212:28
_kJ_@file:///local/ferre/prog/ocaml/fablis/instances/jsoniq/html/script.js:11198:26
_alf_@file:///local/ferre/prog/ocaml/fablis/instances/jsoniq/html/script.js:12129:22
_alj_@file:///local/ferre/prog/ocaml/fablis/instances/jsoniq/html/script.js:12126:46
caml_call1@file:///local/ferre/prog/ocaml/fablis/instances/jsoniq/html/script.js:2212:28
_kJ_@file:///local/ferre/prog/ocaml/fablis/instances/jsoniq/html/script.js:11198:26
_alg_@file:///local/ferre/prog/ocaml/fablis/instances/jsoniq/html/script.js:12128:24
caml_call1@file:///local/ferre/prog/ocaml/fablis/instances/jsoniq/html/script.js:2212:28
_kJ_@file:///local/ferre/prog/ocaml/fablis/instances/jsoniq/html/script.js:11198:26
_alf_@file:///local/ferre/prog/ocaml/fablis/instances/jsoniq/html/script.js:12129:22
_alj_@file:///local/ferre/prog/ocaml/fablis/instances/jsoniq/html/script.js:12126:46
caml_call1@file:///local/ferre/prog/ocaml/fablis/instances/jsoniq/html/script.js:2212:28
_kJ_@file:///local/ferre/prog/ocaml/fablis/instances/jsoniq/html/script.js:11198:26
_alg_@file:///local/ferre/prog/ocaml/fablis/instances/jsoniq/html/script.js:12128:24
caml_call1@file:///local/ferre/prog/ocaml/fablis/instances/jsoniq/html/script.js:2212:28
_kJ_@file:///local/ferre/prog/ocaml/fablis/instances/jsoniq/html/script.js:11198:26
_alf_@file:///local/ferre/prog/ocaml/fablis/instances/jsoniq/html/script.js:12129:22
_mT_@file:///local/ferre/prog/ocaml/fablis/instances/jsoniq/html/script.js:12130:20
_adV_/<@file:///local/ferre/prog/ocaml/fablis/instances/jsoniq/html/script.js:13136:43
caml_call1@file:///local/ferre/prog/ocaml/fablis/instances/jsoniq/html/script.js:2212:28
_zV_/<@file:///local/ferre/prog/ocaml/fablis/instances/jsoniq/html/script.js:18902:40
caml_call1@file:///local/ferre/prog/ocaml/fablis/instances/jsoniq/html/script.js:2212:28
_akD_@file:///local/ferre/prog/ocaml/fablis/instances/jsoniq/html/script.js:12196:37
caml_call1@file:///local/ferre/prog/ocaml/fablis/instances/jsoniq/html/script.js:2212:28
_kK_@file:///local/ferre/prog/ocaml/fablis/instances/jsoniq/html/script.js:11201:54
_akx_/_akA_.onloadend<@file:///local/ferre/prog/ocaml/fablis/instances/jsoniq/html/script.js:12200:26
caml_call1@file:///local/ferre/prog/ocaml/fablis/instances/jsoniq/html/script.js:2212:28
_kU_/<@file:///local/ferre/prog/ocaml/fablis/instances/jsoniq/html/script.js:11231:27
script.js:12524:37

**Comment by hhugo** [comment]
Make sure to build your bytecode with debug info (`-g`) and give `--pretty` to js_of_ocaml.

**Comment by hhugo** [comment]
Are you able to produce a minimal reproduction case ? can you share it (privately maybe) ?

**Comment by sebferre** [comment]
Here is a more readable trace indeed, with -g

TypeError: rev_sort is not a function
Stack trace:
sort@file:///local/ferre/prog/ocaml/fablis/instances/jsoniq/html/script.js:15631:29
eval_flower@file:///local/ferre/prog/ocaml/fablis/instances/jsoniq/html/script.js:15700:29
eval_expr@file:///local/ferre/prog/ocaml/fablis/instances/jsoniq/html/script.js:16018:25
_oQ_/<@file:///local/ferre/prog/ocaml/fablis/instances/jsoniq/html/script.js:19509:21
caml_call3@file:///local/ferre/prog/ocaml/fablis/instances/jsoniq/html/script.js:2216:28
render_place@file:///local/ferre/prog/ocaml/fablis/instances/jsoniq/html/script.js:20947:20
caml_call2@file:///local/ferre/prog/ocaml/fablis/instances/jsoniq/html/script.js:2214:28
render_hist@file:///local/ferre/prog/ocaml/fablis/instances/jsoniq/html/script.js:13013:33
callback_hist@file:///local/ferre/prog/ocaml/fablis/instances/jsoniq/html/script.js:13016:28
caml_call2@file:///local/ferre/prog/ocaml/fablis/instances/jsoniq/html/script.js:2214:28
_pD_/<@file:///local/ferre/prog/ocaml/fablis/instances/jsoniq/html/script.js:20945:58
caml_call1@file:///local/ferre/prog/ocaml/fablis/instances/jsoniq/html/script.js:2212:28
select_suggestion@file:///local/ferre/prog/ocaml/fablis/instances/jsoniq/html/script.js:13565:26
_Cj_/</<@file:///local/ferre/prog/ocaml/fablis/instances/jsoniq/html/script.js:13684:54
caml_call1@file:///local/ferre/prog/ocaml/fablis/instances/jsoniq/html/script.js:2212:28
find_ancestor@file:///local/ferre/prog/ocaml/fablis/instances/jsoniq/html/script.js:12629:23
_E8_@file:///local/ferre/prog/ocaml/fablis/instances/jsoniq/html/script.js:12631:46
caml_call1@file:///local/ferre/prog/ocaml/fablis/instances/jsoniq/html/script.js:2212:28
iter$3@file:///local/ferre/prog/ocaml/fablis/instances/jsoniq/html/script.js:11677:25
_E7_@file:///local/ferre/prog/ocaml/fablis/instances/jsoniq/html/script.js:12633:24
caml_call1@file:///local/ferre/prog/ocaml/fablis/instances/jsoniq/html/script.js:2212:28
iter$3@file:///local/ferre/prog/ocaml/fablis/instances/jsoniq/html/script.js:11677:25
find_ancestor@file:///local/ferre/prog/ocaml/fablis/instances/jsoniq/html/script.js:12634:22
_E8_@file:///local/ferre/prog/ocaml/fablis/instances/jsoniq/html/script.js:12631:46
caml_call1@file:///local/ferre/prog/ocaml/fablis/instances/jsoniq/html/script.js:2212:28
iter$3@file:///local/ferre/prog/ocaml/fablis/instances/jsoniq/html/script.js:11677:25
_E7_@file:///local/ferre/prog/ocaml/fablis/instances/jsoniq/html/script.js:12633:24
caml_call1@file:///local/ferre/prog/ocaml/fablis/instances/jsoniq/html/script.js:2212:28
iter$3@file:///local/ferre/prog/ocaml/fablis/instances/jsoniq/html/script.js:11677:25
find_ancestor@file:///local/ferre/prog/ocaml/fablis/instances/jsoniq/html/script.js:12634:22
_E8_@file:///local/ferre/prog/ocaml/fablis/instances/jsoniq/html/script.js:12631:46
caml_call1@file:///local/ferre/prog/ocaml/fablis/instances/jsoniq/html/script.js:2212:28
iter$3@file:///local/ferre/prog/ocaml/fablis/instances/jsoniq/html/script.js:11677:25
_E7_@file:///local/ferre/prog/ocaml/fablis/instances/jsoniq/html/script.js:12633:24
caml_call1@file:///local/ferre/prog/ocaml/fablis/instances/jsoniq/html/script.js:2212:28
iter$3@file:///local/ferre/prog/ocaml/fablis/instances/jsoniq/html/script.js:11677:25
find_ancestor@file:///local/ferre/prog/ocaml/fablis/instances/jsoniq/html/script.js:12634:22
jquery_ancestor@file:///local/ferre/prog/ocaml/fablis/instances/jsoniq/html/script.js:12635:20
_Cj_/<@file:///local/ferre/prog/ocaml/fablis/instances/jsoniq/html/script.js:13680:43
caml_call1@file:///local/ferre/prog/ocaml/fablis/instances/jsoniq/html/script.js:2212:28
_pO_/<@file:///local/ferre/prog/ocaml/fablis/instances/jsoniq/html/script.js:20568:40
caml_call1@file:///local/ferre/prog/ocaml/fablis/instances/jsoniq/html/script.js:2212:28
_EV_@file:///local/ferre/prog/ocaml/fablis/instances/jsoniq/html/script.js:12704:37
caml_call1@file:///local/ferre/prog/ocaml/fablis/instances/jsoniq/html/script.js:2212:28
case$0@file:///local/ferre/prog/ocaml/fablis/instances/jsoniq/html/script.js:11680:52
_EU_/reader.onloadend<@file:///local/ferre/prog/ocaml/fablis/instances/jsoniq/html/script.js:12708:26
caml_call1@file:///local/ferre/prog/ocaml/fablis/instances/jsoniq/html/script.js:2212:28
handler/<@file:///local/ferre/prog/ocaml/fablis/instances/jsoniq/html/script.js:11710:25
script.js:13028:57

**Comment by sebferre** [comment]
Sorry, I don't have time to build a minimal reproduction case. I tried a few very simple cases
but they don't trigger the error. The code involved is spread over a dozen files, and I don't have
a clue about what is essential or not. 
I can send you privately HTML+CSS+JS if you want to trace it on your side.

**Comment by hhugo** [comment]
I've sent you an email.

**Comment by hhugo** [comment]
I was able to reproduce the issue and I think you should be able to work around it temporarily by passing `--disable inline` to js_of_ocaml

**Comment by sebferre** [comment]
Good to know, thanks. Does this have a significant impact on performance though ?

**Comment by hhugo** [comment]
I'm seeing the issue with the following program.
```ocaml
type t =
  | Empty
  | Node of t

let rec f x =
  match x with
  | Empty -> ()
  | Node next ->
      let _ = List.sort compare [ 1; 2; 3; 4 ] in
      f next

let _ = f (Node Empty)
```

**Comment by hhugo** [comment]
Would you be able to test this branch https://github.com/ocsigen/js_of_ocaml/tree/fix-1007 and confirm that it fixes your issue ?

**Comment by hhugo** [comment]
Fixed by #1009
